File: modules/domain/src/main/java/com/streamfirst/iceberg/hybrid/domain/CommitApproval.java
package com.streamfirst.iceberg.hybrid.domain;

import java.time.Instant;
import lombok.NonNull;

/**
 * Records a single approval or rejection from a region. Simple, orthogonal design - only tracks
 * what happened, not business rules.
 */
public record CommitApproval(
    @NonNull CommitRequest request,
    @NonNull Region region,
    @NonNull ApprovalStatus status,
    @NonNull Instant timestamp,
    String reason // optional reason for rejection
    ) {

  /** Creates an approval record. */
  public static CommitApproval approved(CommitRequest request, Region region) {
    return new CommitApproval(request, region, ApprovalStatus.APPROVED, Instant.now(), null);
  }

  /** Creates a rejection record. */
  public static CommitApproval rejected(CommitRequest request, Region region, String reason) {
    return new CommitApproval(request, region, ApprovalStatus.REJECTED, Instant.now(), reason);
  }

  /** Creates a pending record. */
  public static CommitApproval pending(CommitRequest request, Region region) {
    return new CommitApproval(request, region, ApprovalStatus.PENDING, Instant.now(), null);
  }

  public enum ApprovalStatus {
    APPROVED,
    REJECTED,
    PENDING
  }
}

File: modules/domain/src/main/scala/com/streamfirst/iceberg/hybrid/domain/CommitApproval.scala
package com.streamfirst.iceberg.hybrid.domain

import java.time.Instant

/** Represents approval for a commit request from the global commit gate. Once approved, the commit
  * can proceed and sync events will be created.
  */
final case class CommitApproval(
  requestId: String,
  tableId: TableId,
  commitId: CommitId,
  approvedAt: Instant,
  approvedRegions: List[Region]
)

object CommitApproval:
  def approve(
    request: CommitRequest,
    regions: List[Region]
  ): CommitApproval =
    CommitApproval(
      requestId = request.requestId,
      tableId = request.tableId,
      commitId = request.metadata.commitId,
      approvedAt = Instant.now(),
      approvedRegions = regions
    )
